File: back/src/echoroo/schemas/detection_visualization.py
"""Schemas for detection visualization."""

from __future__ import annotations

import datetime as dt

from pydantic import BaseModel, Field

__all__ = [
    "DetectionTemporalData",
    "SpeciesTemporalData",
    "HourlyDetection",
]


class HourlyDetection(BaseModel):
    """Detection count for a specific date and hour."""

    date: dt.date
    """The date of detections."""

    hour: int = Field(ge=0, le=23)
    """Hour of day (0-23)."""

    count: int
    """Number of detections in this hour."""


class SpeciesTemporalData(BaseModel):
    """Temporal detection data for a single species."""

    scientific_name: str
    """Scientific name of the species."""

    common_name: str | None = None
    """Common name (localized)."""

    total_detections: int
    """Total number of detections for this species."""

    detections: list[HourlyDetection]
    """List of hourly detection counts."""


class DetectionTemporalData(BaseModel):
    """Complete temporal data for all detected species."""

    run_uuid: str
    """Foundation model run UUID."""

    filter_application_uuid: str | None = None
    """Species filter application UUID if filtered."""

    date_range: tuple[dt.date, dt.date] | None = None
    """Date range of detections (min, max)."""

    species: list[SpeciesTemporalData]
    """Temporal data for each species."""

File: front/src/lib/api/detection_visualization.ts
"use client";

import { AxiosInstance } from "axios";
import { z } from "zod";

// ============================================================================
// Schemas
// ============================================================================

export const DetectionTemporalDataPointSchema = z.object({
  date: z.string(),
  hour: z.number().int().min(0).max(23),
  count: z.number().int().min(0),
});

export const SpeciesTemporalDataSchema = z.object({
  scientific_name: z.string(),
  common_name: z.string().nullable().optional(),
  total_detections: z.number().int(),
  detections: z.array(DetectionTemporalDataPointSchema),
});

export const DetectionTemporalResponseSchema = z.object({
  run_uuid: z.string().uuid(),
  filter_application_uuid: z.string().uuid().nullable().optional(),
  species: z.array(SpeciesTemporalDataSchema),
  date_range: z.tuple([z.string(), z.string()]).nullable().optional(),
});

// ============================================================================
// Types
// ============================================================================

export type DetectionTemporalDataPoint = z.infer<
  typeof DetectionTemporalDataPointSchema
>;
export type SpeciesTemporalData = z.infer<typeof SpeciesTemporalDataSchema>;
export type DetectionTemporalResponse = z.infer<
  typeof DetectionTemporalResponseSchema
>;

// ============================================================================
// API Endpoints
// ============================================================================

const DEFAULT_ENDPOINTS = {
  temporal: "/api/v1/detection_visualization/temporal/",
  temporalInference: "/api/v1/detection_visualization/temporal_inference/",
};

// ============================================================================
// API Registration
// ============================================================================

export function registerDetectionVisualizationAPI(instance: AxiosInstance) {
  /**
   * Get temporal detection data for visualization.
   * Returns detection counts grouped by species, date, and hour.
   */
  async function getTemporalData({
    runUuid,
    filterApplicationUuid,
  }: {
    runUuid: string;
    filterApplicationUuid?: string;
  }): Promise<DetectionTemporalResponse> {
    const params: Record<string, string> = {
      run_uuid: runUuid,
    };
    if (filterApplicationUuid) {
      params.filter_application_uuid = filterApplicationUuid;
    }
    const { data } = await instance.get(DEFAULT_ENDPOINTS.temporal, { params });
    return DetectionTemporalResponseSchema.parse(data);
  }

  /**
   * Get temporal inference batch data for visualization.
   * Returns detection counts grouped by date and hour for a specific inference batch.
   */
  async function getInferenceBatchTemporal({
    batchUuid,
  }: {
    batchUuid: string;
  }): Promise<DetectionTemporalResponse> {
    const params: Record<string, string> = {
      batch_uuid: batchUuid,
    };
    const { data } = await instance.get(DEFAULT_ENDPOINTS.temporalInference, { params });
    return DetectionTemporalResponseSchema.parse(data);
  }

  return {
    getTemporalData,
    getInferenceBatchTemporal,
  };
}
